File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> existing = new HashMap<>();
        existing.put(null, null);
        Node dummy = new Node(0);
        Node curr = dummy;
        while (head != null) {
            curr.next = getOrCreateNode(existing, head);
            curr = curr.next;
            existing.put(head, curr);
            curr.random = getOrCreateNode(existing, head.random);
            existing.put(head.random, curr.random);
            head = head.next;
        }
        return dummy.next;
    }

    private Node getOrCreateNode(Map<Node, Node> existing, Node key) {
        if (existing.containsKey(key)) {
            return existing.get(key);
        }
        return new Node(key.val);
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        mapping = {None: None}
        dummy = Node(0)
        curr = dummy
        while head:
            if head in mapping:
                curr.next = mapping[head]
            else:
                curr.next = Node(head.val)
                mapping[head] = curr.next
            curr = curr.next
            if head.random in mapping:
                curr.random = mapping[head.random]
            else:
                curr.random = Node(head.random.val)
                mapping[head.random] = curr.random
            head = head.next
        return dummy.next